src/index/storage.rs

```diff
@@ -66,3 +85,99 @@ impl FileSystemStorage {
         Ok(())
     }
 }
+
+#[derive(Debug)]
+pub struct S3Storage {
+    runtime: Runtime,
+    client: S3Client,
+    bucket: String,
+    key: String,
+}
+
+impl S3Storage {
+    fn new(bucket: &str, key: &str) -> Result<Self> {
+        let runtime = Runtime::new()?;
+
+        let config = runtime.block_on(async {
+            let global_config = aws_config::load_from_env().await;
+            let global_s3 = S3Client::new(&global_config);
+
+            let location = global_s3
+                .get_bucket_location()
+                .bucket(bucket)
+                .send()
+                .await?;
+            let region = location
+                .location_constraint()
+                .map(|c| c.as_str())
+                .unwrap_or("us-east-1")
+                .to_string();
+
+            info!("using S3 bucket {bucket} in region {region}");
+
+            let regional_config = aws_config::from_env()
+                .region(Region::new(region))
+                .load()
+                .await;
+
+            Ok::<_, anyhow::Error>(regional_config)
+        })?;
+        let client = S3Client::new(&config);
+
+        Ok(S3Storage {
+            runtime,
+            client,
+            bucket: bucket.into(),
+            key: key.into(),
+        })
+    }
+
+    fn read(&self) -> Result<Option<Index>> {
+        self.runtime.block_on(async {
+            let result = self
+                .client
+                .get_object()
+                .bucket(&self.bucket)
+                .key(&self.key)
+                .send()
+                .await;
+
+            match result {
+                Ok(response) => {
+                    // FIXME: this buffers the downloaded data into memory before deserializing it,
+                    // as I'm not aware of a way to convert from AsyncRead to Read.
+                    let mut buf = Vec::new();
+                    tokio::io::copy(&mut response.body.into_async_read(), &mut buf).await?;
+                    Ok(Some(Index::deserialize(&mut Cursor::new(buf))?))
+                }
+                Err(err) => {
+                    if let SdkError::ServiceError(service_err) = &err {
+                        if let GetObjectError::NoSuchKey(_) = service_err.err() {
+                            return Ok(None);
+                        }
+                    }
+                    Err(err.into())
+                }
+            }
+        })
+    }
+
+    fn write(&self, index: &Index) -> Result<()> {
+        self.runtime.block_on(async {
+            // FIXME: this buffers the serialized data into memory before sending it, as I'm not
+            // aware of a way to convert from Write to AsyncWrite.
+            let mut buf = Vec::new();
+            index.serialize(&mut Cursor::new(&mut buf))?;
```

Comment:
Cursor feels iffy here? I wouldn't expect `Cursor<Vec<u8>>` to be the right type, Vec already implements Write (and so &mut Vec does too).

Reply:
...not sure why I did that. Reverted to just `&mut buf`.